### DISCUSSION TITLE
adding a button to switch between two views

### DISCUSSION THREAD

**Discussion by Adem-Kri** [eclipse-glsp/glsp#822] [Q&A] [answered]
Hello, I'm new to glsp and I need your assistance.

So, I created a diagram with glsp that shows a model and a theia extension that opens a table with the same data as in the diagram. 
My project is based on the node-theia-json example.

On the two pages, I want to include a button that would let me switch between the two views.
Any idea on how to create this button?

thank you in advance...

**Comment by planger** [comment, marked as answer]
Hi,

I think, I'd implement that via Theia's opener service. Both buttons invoke `open` on click with opener options for which the other view is registered. With that you can achieve that those views would be activated and brought to front, whether they are currently already opened or not. Optionally you could extend the opener options so that the views would also update their selection as a result.

## Prerequisite

There needs to be an [open handler](https://github.com/eclipse-theia/theia/blob/master/packages/core/src/browser/widget-open-handler.ts) for each view that reacts to the open-event for a specific type of URI / opener options (see `canHandle(URI, OpenerOptions)` of your `OpenHandler`).

## Open Event From GLSP

The button in GLSP can be implemented with a [UI Extension](https://www.eclipse.org/glsp/documentation/ui-extensions/) or any other way. Likely, you'd send a [custom GLSP action](https://www.eclipse.org/glsp/documentation/actionhandler/) as a result. In your Theia integration, you'd then register an action handler for this action, which in turn invokes a Theia command.

## Theia command (From GLSP or your Theia View)

The Theia command for invoking the open handler is very similar for both ends, from GLSP or from your Theia view. Eventually it needs to call `open(this.openerService, uri, openerOptions)` so that your open handlers (see prerequsiites) are triggered.
See the [`ExampleNavigationCommandContribution`](https://github.com/eclipse-glsp/glsp-theia-integration/blob/master/examples/workflow-theia/src/browser/external-navigation-example/external-navigation-example.ts) as an example. In this example, we invoke the open handler of the workflow GLSP editor with an additional opener option (`selection: { name: 'Push' } }`, which then eventually leads to selecting the `Push` diagram element. So this example is more or less equivalent to what you need in the direction from your Theia view to GLSP. The other direction, however, is pretty much the same, except that for this URI and opener option another open handler is registered that just activates your Theia view.

I hope this gives some hints into the right direction!
